torch/distributed/_tensor/dispatch.py

```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+            "float",
+            "double",
+        ]
+        return arg_type in _python_primitive_types
+
     if mesh is not None and mesh.get_coordinate() is None:
-        # if we are on a non-participating device, we simply return
-        # an empty tensor for now.
-        # TODO: what if the op returns a non-tensor value, what if
-        # the op returns a list of tensors, we need to figure out
-        # a consistent way to handle that, and also need to figure
-        # out if we should communicate the result to non-participating
-        # ranks (i.e. a.sum() -> scalar, maybe we should set to 0)
-        local_results = torch.tensor([])
+        # For a non-participating device, we do:
+        # 1. if the return type is scalar, all gather the local result
+        # from participating devices, and reduce on the list of results.
+        # For bool type, we by default use AND to reduce;
+        # For numeric type, we by default use ADD to reduce;
+        # We can extend this logic for specific ops if necessary.
+        # 2. if the return type is Tensor, return empty tensor with
+        # correct dtype.
+        # 3. if the return type is List[Tensor], return a list of empty
+        # tensors with correct dtype.
+
+        def _default_tensor(spec: Optional[DTensorSpec]):
+            assert spec is not None, (
+                f"expect DTensorSpec object but got {spec}"
+            )
+            if spec.tensor_meta is not None:
+                shape = spec.tensor_meta.shape
+                dtype = spec.tensor_meta.dtype
+                if len(shape) == 0:
+                    # scalar tensor
+                    return torch.zeros(
+                        (),
+                        dtype=dtype,
+                    )
+                else:
+                    # non-scalar tensor
+                    return torch.tensor(
+                        [],
+                        dtype=dtype,
+                    )
+            else:
+                raise RuntimeError(
+                    f"{spec} has no tensor metadata."
+                )
+
+        def _default_ret_value(arg):
+            type_str = str(arg.type)
+            if type_str == "()":  # used in aten function schema for None return value
+                return None
+
+            if _arg_type_is_primitive(type_str):
+                obj_list = [None for _ in range(dist.get_world_size())]
+                dist.all_gather_object(obj_list, None)
+                obj_list = list(filter(lambda x: x is not None, obj_list))
+                # TODO: this nested if-else statement is due to MYPY's
+                # issue on inferring reduce op's type. We have to explicitly
+                # type the lambda to make MYPY happy. But we better find a
+                # smarter way to do it.
+                if type_str == "bool":
+                    return functools.reduce(lambda x, y: x and y, obj_list, True)  # type: ignore[arg-type]
+                elif (
```

Comment:
I think a simple "else" branch would be sufficient here, as you are already in `_arg_type_is_primitive==True` condition